### PR TITLE
Load JSON configuration on ConfigBuilder

### PR DESCRIPTION
**Pull request by mdumandag**
With this pr, JSON configuration is loaded in the ConfigBuilder instead of just importing and loading them when they needed. 

There are 2 advantages,
1- Internal fields like `listenerConfigs`,  `dataSerializableFactoryConfigs`,  `sslOptionsFactoryConfig` are no longer being exposed. It is confusing when both `portableFactories`(used to programmatically set the factories) and `portableFactoryConfigs`(used internally to import portable factories defined on JSON config file) are in the public API.

2- Invalid JSON configuration files are detected at the ConfigBuilder so we fail fast.

### CODE REVIEW

File: src/config/ConfigBuilder.ts [hunk 1]
```diff
@@ -185,8 +194,10 @@ export class ConfigBuilder {
     private handleListeners(jsonObject: any): void {
         const listenersArray = tryGetArray(jsonObject);
         for (const index in listenersArray) {
-            const listenerConfig = listenersArray[index];
-            this.clientConfig.listenerConfigs.push(this.parseImportConfig(listenerConfig));
+            const listenerObj = listenersArray[index];
+            const importConfig = this.parseImportConfig(listenerObj);
+            const listener = loadNameFromPath(importConfig.path, importConfig.exportedName);
+            this.clientConfig.listeners.lifecycle.push(listener);
```
[the remark this answers is not in the document]
Comment (mdumandag) [reply src/config/ConfigBuilder.ts]: done

File: test/declarative_config/ConfigBuilderFailFastTest.js [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+            '           "factory": {' +
+            '               "exportedName": "invalid"' +
+            '           }' +
+            '       }' +
+            '   }' +
+            '}');
+    });
+
+    function testConfig(jsonConfigString) {
+        fs.writeFileSync(JSON_LOCATION, jsonConfigString);
```
[the remark this answers is not in the document]
Comment (mdumandag) [reply test/declarative_config/ConfigBuilderFailFastTest.js]: I changed the tests, now parsed configurations are directly passed to configBuilder.loadedJson.

File: test/LifecycleServiceTest.js [hunk 3]
```diff
@@ -123,5 +103,58 @@ describe('LifecycleService', function () {
             client.shutdown();
             done();
         });
-    })
+    });
+
+    describe('DeclarativeConfiguration', function () {
```
[the remark this answers is not in the document]
Comment (mdumandag) [reply test/LifecycleServiceTest.js]: Right, there is a test on `ConfigBuilderTest#listenerConfig` that checks whether or not listeners are loaded from json configuration, so I removed this test

File: test/LifecycleServiceTest.js [hunk 4]
```diff
@@ -57,28 +56,6 @@ describe('LifecycleService', function () {
         });
     });
 
-    it('client should emit starting, started, shuttingDown and shutdown events in order (via import config)', function (done) {
```
Comment (mustafaiman) [inline test/LifecycleServiceTest.js]: Is there an alternative to this test somewhere else? Why is it removed?

Comment (mdumandag) [reply test/LifecycleServiceTest.js]: Loading lifecycle listeners from the config file test is covered in `ConfigBuilderTest#listenerConfig` and emitting lifecycle events test is covered `LifecycleServiceTest#client should emit starting, started, shuttingDown and shutdown events in order`.

**Comment by puzpuzpuz** [comment]
@mdumandag considering that declarative configuration support was removed in 4.0, is it OK to close this PR? Or you think it's worth it to update it and merge into 3.12.x?

**Comment by mdumandag** [comment]
We can close this. This one was mainly to get rid of confusing public fields that make no sense to the user. We don't have this problem anymore.

**Pull request closed without merging**